automotiveApp/build.gradle.kts

```diff
@@ -141,11 +141,15 @@ dependencies {
     implementation(libs.compose.compiler)
     implementation(libs.lifecycle.runtime.compose)
 
+    val excludeAndroidxDataStore = Action<ExternalModuleDependency> {
+        // Crashlytics and PerfMon depend on datastore v1.0 but we're using v1.1
+        exclude(group = "androidx.datastore", module = "datastore-preferences")
+    }
     implementation(platform(libs.firebase.bom))
     implementation(libs.google.services)
-    implementation(libs.firebase.crashlytics)
+    implementation(libs.firebase.crashlytics, excludeAndroidxDataStore)
```

Comment:
Shouldn't Gradle resolve the latest version of the dependency no matter what? https://docs.gradle.org/current/userguide/dependency_resolution.html#sub:resolution-strategy

And 1.1.0-alpha4 > 1.0.0?

Reply:
@martinbonnin That's what I thought. But for some reason I'm getting this:

```
> Task :androidApp:validateSigningDebug UP-TO-DATE

FAILURE: Build completed with 8 failures.

1: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':androidApp:checkDebugDuplicateClasses'.
> Could not resolve all files for configuration ':androidApp:debugRuntimeClasspath'.
   > Failed to transform datastore-preferences-1.1.0-alpha04.aar (androidx.datastore:datastore-preferences:1.1.0-alpha04) to match attributes {artifactType=enumerated-runtime-classes, org.gradle.status=release, org.gradle.usage=java-runtime}.
      > Could not find datastore-preferences-1.1.0-alpha04.aar (androidx.datastore:datastore-preferences:1.1.0-alpha04).
        Searched in the following locations:
            https://dl.google.com/dl/android/maven2/androidx/datastore/datastore-preferences/1.1.0-alpha04/datastore-preferences-1.1.0-alpha04.aar
```

See all the 8 failures [here](https://gist.github.com/thatfiredev/4a6f6bb36e615fbe8fc0d65a8f1a74ee)

Reply:
Now that I had a closer look at the logs, it seems like Gradle is looking for an aar while [datastore-preferences:1.1.0-alpha04](https://maven.google.com/web/index.html?q=datastore#androidx.datastore:datastore-preferences:1.1.0-alpha04) doesn't have one 🤔  ([alpha01](https://maven.google.com/web/index.html?q=datastore#androidx.datastore:datastore-preferences:1.1.0-alpha01) had it though)

Reply:
Ah yes good catch! Looks like [firebase-session:1.0.0](https://maven.google.com/web/index.html?q=firebase-session#com.google.firebase:firebase-sessions:1.0.0) is asking for `.aar`. From the pom file:

```
    <dependency>
      <groupId>androidx.datastore</groupId>
      <artifactId>datastore-preferences</artifactId>
      <version>1.0.0</version>
      <scope>compile</scope>
      <type>aar</type>
    </dependency>
```

Reply:
Not 100% sure what the good fix is. if it's just a packaging issue, I guess excluding is fine. We'll have to remember to remove the exclude later down the road to enable transitive dependencies resolution again.

Reply:
Should we merge as it is? Then remove the exclude once the a version of datastore with aar comes out?

Reply:
Yea, let's do this 🚀 Thanks!

Reply:
For the record, another fix would be to patch the `firebase-sessions` metadata on the fly with this:

```kotlin
dependencies {
    components {
        withModule("com.google.firebase:firebase-sessions") {
            allVariants {
                withDependencies {
                    // Remove the aar dependency
                    removeIf {
                        (it.group == "androidx.datastore") && (it.name == "datastore-preferences")
                    }
                    // And add a regular .jar dependency
                    add("androidx.datastore:datastore-preferences:1.1.0-alpha04")
                }
            }
        }
    }
}
```

If you have control over `firebase-sessions`, it'd be nice to publish a new version that plays more nicely with `datastore-preferences` (i.e. works with plain .jar dependency) and push that to the BOM

Reply:
> If you have control over firebase-sessions, it'd be nice to publish a new version that plays more nicely with datastore-preferences (i.e. works with plain .jar dependency) and push that to the BOM

Given that none of the Firebase Android SDKs have support for KMP, I doubt that it would get prioritized.

But that library is actually open source, in case anyone has a solution in mind and wants to send a PR:
https://github.com/firebase/firebase-android-sdk/tree/master/firebase-sessions

Reply:
> Given that none of the Firebase Android SDKs have support for KMP

I think it doesn't need KMP? Anyone doing the below in a Android, non-KMP project, will bump into the conflict:

```
dependencies {
  // incompatible dependencies
  implementation("androidx.datastore:datastore-preferences:1.1.0-alpha04")
  implementation("com.google.firebase:firebase-sessions:1.0.0")
}
```

A fix is to depend on the `-android` dependency explicitly:

```
dependencies {
  // this works
  implementation("androidx.datastore:datastore-preferences-android:1.1.0-alpha04")
  implementation("com.google.firebase:firebase-sessions:1.0.0")
}
```

But not sure how familiar the Android users will be with that?

Reply:
I opened https://github.com/firebase/firebase-android-sdk/issues/5176 as a follow up

Reply:
Thanks Martin, we'll continue the discussion there.